src/moscot/_utils.py

```diff
@@ -1,20 +1,56 @@
 # adapted from CellRank
 """Module used to parallelize model fitting."""
 
-from typing import Any, Union, Callable, Optional, Sequence
+from typing import Any, Tuple, Union, Mapping, Callable, Optional, Sequence, TYPE_CHECKING
 from threading import Thread
 from multiprocessing import Manager, cpu_count
 
 from numba import njit
 from scipy.sparse import issparse, spmatrix
+import wrapt
 import joblib as jl
 
 import numpy as np
 
+from moscot.problems.base._compound_problem import B, K
+
+if TYPE_CHECKING:
+    from moscot.problems.base import BaseProblem, BaseCompoundProblem  # type: ignore[attr-defined]
+
 from moscot._types import ArrayLike
 
 jit_kwargs = {"nogil": True, "cache": True, "fastmath": True}
 
+__all__ = ["require_prepare", "require_solution", "parallelize"]
+
+
+# TODO(michalk8): refactor using stage
+@wrapt.decorator
+def require_solution(
+    wrapped: Callable[[Any], Any], instance: "BaseProblem", args: Tuple[Any, ...], kwargs: Mapping[str, Any]
+) -> Any:
+    """Check whether problem has been solved."""
+    from moscot.problems.base import OTProblem, BaseCompoundProblem  # type: ignore[attr-defined]
+
+    if isinstance(instance, OTProblem) and instance.solution is None:
+        raise RuntimeError("Run `.solve()` first.")
+    if isinstance(instance, BaseCompoundProblem) and instance.solutions is None:
+        raise RuntimeError("Run `.solve()` first.")
```

Comment:
```suggestion
    if isinstance(instance, (OTProblem, BaseCompoundProblem)) and instance.solution is None:
        raise RuntimeError("Run `.solve()` first.")
```

Reply:
this should also work

Reply:
can't see why tbh. once we check for `instance.solution` once for `instance.solutions`

Reply:
ahh good points missed the `solution`/`solutions` apologies